drivers/spi/spi_sam0.c

```diff
@@ -0,0 +1,432 @@
+	u32_t pm_apbcmask;
+	u16_t gclk_clkctrl_id;
+	struct soc_gpio_pin pin_miso;
+	struct soc_gpio_pin pin_mosi;
+	struct soc_gpio_pin pin_sck;
+};
+
+/* Device run time data */
+struct spi_sam0_data {
+	struct spi_context ctx;
+};
+
+static void wait_synchronization(SercomSpi *regs)
+{
+	while ((regs->SYNCBUSY.reg & SERCOM_SPI_SYNCBUSY_MASK) != 0) {
+	}
+}
+
+static int spi_sam0_configure(struct spi_config *config)
+{
+	const struct spi_sam0_config *cfg = config->dev->config->config_info;
+	SercomSpi *regs = cfg->regs;
+	SERCOM_SPI_CTRLA_Type ctrla = {.reg = 0};
+	SERCOM_SPI_CTRLB_Type ctrlb = {.reg = 0};
+	int div;
+
+	if (SPI_OP_MODE_GET(config->operation) != SPI_OP_MODE_MASTER) {
+		/* Slave mode is not implemented. */
+		return -ENOTSUP;
+	}
+
+	ctrla.bit.MODE = SERCOM_SPI_CTRLA_MODE_SPI_MASTER_Val;
+
+	if ((config->operation & SPI_TRANSFER_LSB) != 0) {
+		ctrla.bit.DORD = 1;
+	}
+
+	if ((config->operation & SPI_MODE_CPOL) != 0) {
+		ctrla.bit.CPOL = 1;
+	}
+
+	if ((config->operation & SPI_MODE_CPHA) != 0) {
+		ctrla.bit.CPHA = 1;
+	}
+
+	/* MOSI on PAD2, SCK on PAD3 */
+	ctrla.bit.DOPO = 1;
+
+	if ((config->operation & SPI_MODE_LOOP) != 0) {
+		/* Put MISO on the same pin as MOSI */
+		ctrla.bit.DIPO = 2;
+	} else {
+		ctrla.bit.DIPO = 0;
+	}
+
+	ctrla.bit.ENABLE = 1;
+	ctrlb.bit.RXEN = 1;
+
+	if (SPI_WORD_SIZE_GET(config->operation) != 8) {
+		return -ENOTSUP;
+	}
+
+	/* 8 bits per transfer */
+	ctrlb.bit.CHSIZE = 0;
+
+	/* Use the requested or next higest possible frequency */
+	div = (SOC_ATMEL_SAM0_GCLK0_FREQ_HZ / config->frequency) / 2 - 1;
+	div = max(0, min(UINT8_MAX, div));
+
+	/* Update the configuration only if it has changed */
+	if (regs->CTRLA.reg != ctrla.reg || regs->CTRLB.reg != ctrlb.reg ||
+	    regs->BAUD.reg != div) {
+		regs->CTRLA.bit.ENABLE = 0;
+		wait_synchronization(regs);
+
+		regs->CTRLB = ctrlb;
+		wait_synchronization(regs);
+		regs->BAUD.reg = div;
+		wait_synchronization(regs);
+		regs->CTRLA = ctrla;
+		wait_synchronization(regs);
+	}
+
+	return 0;
+}
+
+static bool spi_sam0_transfer_ongoing(struct spi_sam0_data *data)
+{
+	return spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx);
+}
+
+static void spi_sam0_shift_master(SercomSpi *regs, struct spi_sam0_data *data)
+{
+	u8_t tx;
+	u8_t rx;
+
+	if (spi_context_tx_on(&data->ctx)) {
+		tx = *(u8_t *)(data->ctx.tx_buf);
+	} else {
+		tx = 0;
+	}
+
+	while (!regs->INTFLAG.bit.DRE) {
+	}
+
+	regs->DATA.reg = tx;
+	spi_context_update_tx(&data->ctx, 1, 1);
+
+	while (!regs->INTFLAG.bit.RXC) {
+	}
+
+	rx = regs->DATA.reg;
+
+	if (spi_context_rx_on(&data->ctx)) {
+		*data->ctx.rx_buf = rx;
+		spi_context_update_rx(&data->ctx, 1, 1);
+	}
+}
+
+/* Fast path that transmits a buf */
+static void spi_sam0_fast_tx(SercomSpi *regs, const struct spi_buf *tx_buf)
+{
+	const u8_t *p = tx_buf->buf;
+	const u8_t *pend = tx_buf->buf + tx_buf->len;
+	u8_t ch;
+
+	while (p != pend) {
+		ch = *p++;
+		while (!regs->INTFLAG.bit.DRE) {
+		}
+		regs->DATA.reg = ch;
```

Comment:
and before this line as well. Actually, apply this comment everywhere: it's ugly to get a statement glued to a previous line with }

Reply:
Done globally.